### DISCUSSION TITLE
Searching for neighbors within a certain Euclidean distance

### DISCUSSION THREAD

**Discussion by SaVoAMP** [networkx/networkx#7464] [Q&A] [answered]
Hey,

I have built a graphene sheet with NetworkX, as you can see here:
![image](https://github.com/networkx/networkx/assets/95933072/400d63f1-f3a7-4d71-bd68-7f54fafac5b6)
The solid black lines represent my bonds and have a bond length as an edge attribute. The nodes are the carbon atoms and have both the element and the position as node attributes. These strange-looking dashed lines correspond to periodic boundary conditions – meaning if you go out from the bottom, you come back in from the top, and the same applies to the left and right.

My problem is as follows:
I want to select a specific atom – for example, C5 – and from there, find all neighbors within a certain distance.

I attempted to solve this problem using the Dijkstra algorithm and obtained the following result:
![image](https://github.com/networkx/networkx/assets/95933072/6477994e-9c57-4e59-8723-b41f7f6a5c3e)
Thus, Dijkstra correctly found all paths that are in total smaller than a certain 'max_distance'. However, for my specific problem, I would like to find all neighbors within a radius of 'max_distance' around the atom C5 (i.e., according to the Euclidean distance). In this specific example, atoms C1, C17, C35, C39, C25, and C9 should also be found.

Is there a graph-theoretical algorithm for this, and if so, have you implemented it in NetworkX? I couldn't find anything. I have managed to solve the problem (at least partially) by additionally using a KDTree, and I have seen that you also use KDTrees in NetworkX. However, what I find inconvenient is that I now have both a graph-theoretical approach and a KDTree approach, essentially having to do everything twice. It was very easy to account for periodic boundary conditions using the graph-theoretical approach, but with the KDTree, it is very cumbersome to find neighbors within a certain range considering periodic boundary conditions.

I know I can also solve the problem by iterating over all atoms and calculating the Euclidean distance to atom C5 each time, but I hope there are more efficient methods (such as the KDTree).

Thank you very much for your help!

**Comment by dschult** [comment, marked as answer]
Using Euclidean Distance assumes the nodes have positions in some space (here it looks 2D, but I suspect in real world chemicals it is 3D). So graph theoretical algorithms will need to be extended to solve that problem. Using something like dijkstra to traverse the network while tracking the euclidean distance as well might get you somewhere. The code for dijkstra is pretty short and understandable, but you would need to figure out how to make decisions when checking the next edge if there are two measures you are minimizing. 

As for handling the periodic boundaries, I would suggest copying the network in each direction from the node in question (C5) and then dropping periodic boundaries on the second pass through the boundary. Then only allow starting the algorithm in the middle copy -- but collect valid nodes in any copy saving only the node name, not which copy it appears in.  Perhaps this approach with a KDTree will give what you want? I think that ignores the network entirely -- but that may be what you want/need.

**Reply by SaVoAMP** [reply]
Alright, thank you very much for the quick response!
